### ISSUE TITLE
ModelCheckpoint save ckpt before Validation on steps

### ISSUE THREAD

**Issue by tshu-w** [Lightning-AI/lightning#14431]
## 🐛 Bug

The `ModelCheckpoint` check whether to save ckpt before `Trainer` `Validation` if val and save checkpoints on steps.

### To Reproduce

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks import ModelCheckpoint


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    model_checkpoint = ModelCheckpoint(monitor="valid_loss", every_n_train_steps=2)
    trainer = Trainer(
        accelerator="auto",
        default_root_dir=os.getcwd(),
        log_every_n_steps=1,
        check_val_every_n_epoch=None,
        val_check_interval=2,
        num_sanity_val_steps=0,
        max_steps=10,
        enable_model_summary=False,
        callbacks=[model_checkpoint],
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)


if __name__ == "__main__":
    run()
```

### Expected behavior

checkpoint save successfully at step 2 rather than complain:
```
/opt/homebrew/Caskroom/miniconda/base/envs/pl/lib/python3.9/site-packages/pytorch_lightning/callbacks/model_checkpoint.py:381: UserWarning: `ModelCheckpoint(monitor='valid_loss')` could not find the monitored key in the returned metrics: ['train_loss', 'epoch', 'step']. HINT: Did you call `log('valid_loss', value)` in the `LightningModule`?
  warning_cache.warn(m)
```

### Environment

<details>
  <summary>Details</summary>
    * CUDA:
	- GPU:               None
	- available:         False
	- version:           None
* Lightning:
	- lightning:         2022.8.21
	- lightning-cloud:   0.5.3
	- pytorch-lightning: 1.7.3
	- torch:             1.12.1
	- torchmetrics:      0.9.2
	- torchtext:         0.13.1
	- torchvision:       0.13.0
* Packages:
	- absl-py:           1.2.0
	- aiobotocore:       2.3.4
	- aiohttp:           3.8.1
	- aioitertools:      0.10.0
	- aiosignal:         1.2.0
	- alabaster:         0.7.12
	- alembic:           1.8.1
	- antlr4-python3-runtime: 4.9.3
	- anyio:             3.6.1
	- appnope:           0.1.3
	- argon2-cffi:       21.3.0
	- argon2-cffi-bindings: 21.2.0
	- asgiref:           3.5.2
	- asttokens:         2.0.8
	- async-generator:   1.10
	- async-timeout:     4.0.2
	- attrs:             22.1.0
	- babel:             2.10.3
	- backcall:          0.2.0
	- beautifulsoup4:    4.11.1
	- black:             22.6.0
	- bleach:            5.0.1
	- boto3:             1.24.56
	- botocore:          1.24.21
	- bravado:           11.0.3
	- bravado-core:      5.17.0
	- cachetools:        5.2.0
	- certifi:           2022.6.15
	- cffi:              1.15.1
	- cfgv:              3.3.1
	- charset-normalizer: 2.1.1
	- click:             8.1.3
	- cloudpickle:       2.1.0
	- codecov:           2.1.12
	- coloredlogs:       15.0.1
	- comet-ml:          3.31.7
	- commonmark:        0.9.1
	- configobj:         5.0.6
	- coverage:          6.4.2
	- croniter:          1.3.5
	- curio:             1.5
	- cycler:            0.11.0
	- databricks-cli:    0.17.2
	- debugpy:           1.6.3
	- decorator:         5.1.1
	- deepdiff:          5.8.1
	- defusedxml:        0.7.1
	- distlib:           0.3.5
	- dnspython:         2.2.1
	- docker:            5.0.3
	- docker-pycreds:    0.4.0
	- docstring-parser:  0.14.1
	- docutils:          0.19
	- dulwich:           0.20.45
	- email-validator:   1.2.1
	- entrypoints:       0.4
	- everett:           3.0.0
	- executing:         0.10.0
	- fastapi:           0.79.0
	- fastjsonschema:    2.16.1
	- filelock:          3.8.0
	- flask:             2.2.2
	- flatbuffers:       2.0
	- fonttools:         4.36.0
	- frozenlist:        1.3.1
	- fsspec:            2022.5.0
	- future:            0.18.2
	- gcsfs:             2022.5.0
	- gitdb:             4.0.9
	- gitpython:         3.1.27
	- google-api-core:   2.8.2
	- google-auth:       2.11.0
	- google-auth-oauthlib: 0.4.6
	- google-cloud-core: 2.3.2
	- google-cloud-storage: 2.5.0
	- google-crc32c:     1.3.0
	- google-resumable-media: 2.3.3
	- googleapis-common-protos: 1.56.4
	- grpcio:            1.47.0
	- gunicorn:          20.1.0
	- gym:               0.24.1
	- gym-notices:       0.0.8
	- h11:               0.13.0
	- hjson:             3.1.0
	- httptools:         0.4.0
	- humanfriendly:     10.0
	- hydra-core:        1.2.0
	- identify:          2.5.3
	- idna:              3.3
	- imageio:           2.21.1
	- imagesize:         1.4.1
	- importlib-metadata: 4.12.0
	- iniconfig:         1.1.1
	- ipykernel:         6.15.1
	- ipyparallel:       8.4.1
	- ipython:           8.4.0
	- ipython-genutils:  0.2.0
	- ipywidgets:        8.0.1
	- itsdangerous:      2.1.2
	- jedi:              0.18.1
	- jinja2:            3.1.2
	- jmespath:          1.0.1
	- joblib:            1.1.0
	- jsonargparse:      4.12.0
	- jsonpointer:       2.3
	- jsonref:           0.2
	- jsonschema:        3.2.0
	- jupyter-client:    7.3.4
	- jupyter-core:      4.11.1
	- jupyterlab-pygments: 0.2.2
	- jupyterlab-widgets: 3.0.2
	- kiwisolver:        1.4.4
	- lightning:         2022.8.21
	- lightning-cloud:   0.5.3
	- lxml:              4.9.1
	- mako:              1.2.1
	- markdown:          3.4.1
	- markupsafe:        2.1.1
	- matplotlib:        3.5.2
	- matplotlib-inline: 0.1.6
	- mistune:           0.8.4
	- mlflow:            1.27.0
	- monotonic:         1.6
	- mpmath:            1.2.1
	- msgpack:           1.0.4
	- multidict:         6.0.2
	- mypy:              0.971
	- mypy-extensions:   0.4.3
	- nbclient:          0.6.6
	- nbconvert:         6.5.3
	- nbformat:          5.4.0
	- neptune-client:    0.16.3
	- nest-asyncio:      1.5.5
	- ninja:             1.10.2.3
	- nodeenv:           1.7.0
	- notebook:          6.4.12
	- numpy:             1.23.0
	- nvidia-ml-py3:     7.352.0
	- oauthlib:          3.2.0
	- omegaconf:         2.2.3
	- onnxruntime:       1.12.1
	- ordered-set:       4.1.0
	- orjson:            3.7.12
	- outcome:           1.2.0
	- packaging:         21.3
	- pandas:            1.4.3
	- pandocfilters:     1.5.0
	- parso:             0.8.3
	- pathspec:          0.9.0
	- pathtools:         0.1.2
	- pexpect:           4.8.0
	- pickleshare:       0.7.5
	- pillow:            9.2.0
	- pip:               22.1.2
	- platformdirs:      2.5.2
	- pluggy:            1.0.0
	- pre-commit:        2.20.0
	- prometheus-client: 0.14.1
	- prometheus-flask-exporter: 0.20.3
	- promise:           2.3
	- prompt-toolkit:    3.0.30
	- protobuf:          3.19.4
	- psutil:            5.9.1
	- ptyprocess:        0.7.0
	- pure-eval:         0.2.2
	- py:                1.11.0
	- py-cpuinfo:        8.0.0
	- pyasn1:            0.4.8
	- pyasn1-modules:    0.2.8
	- pycparser:         2.21
	- pydantic:          1.9.2
	- pydeprecate:       0.3.2
	- pygame:            2.1.0
	- pygments:          2.13.0
	- pyjwt:             2.4.0
	- pyparsing:         3.0.9
	- pyrsistent:        0.18.1
	- pytest:            7.0.1
	- pytest-asyncio:    0.19.0
	- pytest-cov:        3.0.0
	- pytest-forked:     1.4.0
	- pytest-rerunfailures: 10.2
	- python-dateutil:   2.8.2
	- python-dotenv:     0.20.0
	- python-multipart:  0.0.5
	- pytorch-lightning: 1.7.3
	- pytz:              2022.2.1
	- pyyaml:            6.0
	- pyzmq:             23.2.1
	- qtconsole:         5.3.1
	- qtpy:              2.2.0
	- querystring-parser: 1.2.4
	- requests:          2.28.1
	- requests-oauthlib: 1.3.1
	- requests-toolbelt: 0.9.1
	- rfc3987:           1.3.8
	- rich:              12.5.1
	- rsa:               4.9
	- s3fs:              2022.5.0
	- s3transfer:        0.6.0
	- scikit-learn:      1.1.2
	- scipy:             1.9.0
	- semantic-version:  2.10.0
	- send2trash:        1.8.0
	- sentry-sdk:        1.9.5
	- setproctitle:      1.3.2
	- setuptools:        61.2.0
	- shortuuid:         1.0.9
	- simplejson:        3.17.6
	- six:               1.16.0
	- smmap:             5.0.0
	- sniffio:           1.2.0
	- snowballstemmer:   2.2.0
	- sortedcontainers:  2.4.0
	- soupsieve:         2.3.2.post1
	- sphinx:            5.1.1
	- sphinxcontrib-applehelp: 1.0.2
	- sphinxcontrib-devhelp: 1.0.2
	- sphinxcontrib-htmlhelp: 2.0.0
	- sphinxcontrib-jsmath: 1.0.1
	- sphinxcontrib-qthelp: 1.0.3
	- sphinxcontrib-serializinghtml: 1.1.5
	- sqlalchemy:        1.4.40
	- sqlparse:          0.4.2
	- stack-data:        0.4.0
	- starlette:         0.19.1
	- starsessions:      2.0.1
	- strict-rfc3339:    0.7
	- swagger-spec-validator: 2.7.4
	- sympy:             1.10.1
	- tabulate:          0.8.10
	- tensorboard:       2.10.0
	- tensorboard-data-server: 0.6.1
	- tensorboard-plugin-wit: 1.8.1
	- terminado:         0.15.0
	- test-tube:         0.7.5
	- testpath:          0.6.0
	- threadpoolctl:     3.1.0
	- tinycss2:          1.1.1
	- toml:              0.10.2
	- tomli:             2.0.1
	- torch:             1.12.1
	- torchmetrics:      0.9.2
	- torchtext:         0.13.1
	- torchvision:       0.13.0
	- tornado:           6.2
	- tqdm:              4.64.0
	- traitlets:         5.3.0
	- trio:              0.21.0
	- typing-extensions: 4.3.0
	- ujson:             5.4.0
	- urllib3:           1.26.11
	- uvicorn:           0.17.6
	- uvloop:            0.16.0
	- virtualenv:        20.16.3
	- wandb:             0.13.1
	- watchgod:          0.8.2
	- wcwidth:           0.2.5
	- webcolors:         1.12
	- webencodings:      0.5.1
	- websocket-client:  1.3.3
	- websockets:        10.3
	- werkzeug:          2.2.2
	- wheel:             0.37.1
	- widgetsnbextension: 4.0.2
	- wrapt:             1.14.1
	- wurlitzer:         3.0.2
	- yarl:              1.8.1
	- zipp:              3.8.1
* System:
	- OS:                Darwin
	- architecture:
		- 64bit
		- 
	- processor:         arm
	- python:            3.9.12
	- version:           Darwin Kernel Version 21.6.0: Wed Aug 10 14:28:23 PDT 2022; root:xnu-8020.141.5~2/RELEASE_ARM64_T6000
</details>

### Additional context

Full Logs:
```
pl λ python bug_report_model.py
GPU available: True (mps), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
/opt/homebrew/Caskroom/miniconda/base/envs/pl/lib/python3.9/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:225: PossibleUserWarning: The dataloader, train_dataloader, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 8 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
/opt/homebrew/Caskroom/miniconda/base/envs/pl/lib/python3.9/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:225: PossibleUserWarning: The dataloader, val_dataloader 0, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 8 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
Epoch 0:   0%|▌                                                                                                                                          | 2/544 [00:00<00:14, 38.33it/s, loss=-0.0672, v_num=3]/opt/homebrew/Caskroom/miniconda/base/envs/pl/lib/python3.9/site-packages/pytorch_lightning/callbacks/model_checkpoint.py:381: UserWarning: `ModelCheckpoint(monitor='valid_loss')` could not find the monitored key in the returned metrics: ['train_loss', 'epoch', 'step']. HINT: Did you call `log('valid_loss', value)` in the `LightningModule`?
  warning_cache.warn(m)
Epoch 0:  12%|█████████████████▎                                                                                                                        | 68/544 [00:00<00:01, 373.56it/s, loss=0.0935, v_num=3]`Trainer.fit` stopped: `max_steps=4` reached.
Epoch 0:  12%|█████████████████▎                                                                                                                        | 68/544 [00:00<00:01, 372.78it/s, loss=0.0935, v_num=3]
/opt/homebrew/Caskroom/miniconda/base/envs/pl/lib/python3.9/site-packages/pytorch_lightning/trainer/connectors/data_connector.py:225: PossibleUserWarning: The dataloader, test_dataloader 0, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 8 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
Testing DataLoader 0: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 32/32 [00:00<00:00, 676.88it/s]
```

The `ModelCheckpoint._save_topk_checkpoint` is run at `on_train_batch_end` in `TrainingEpochLoop.advance`, while `validating` is run at `on_advance_end`
https://github.com/Lightning-AI/lightning/blob/1a3fe395715b2b9eb3c80647b5e0eb490047be8d/src/pytorch_lightning/loops/epoch/training_epoch_loop.py#L233-L256

cc @borda @carmocca @awaelchli @ninginthecloud @jjenniferdai @rohitgr7

**Comment by tshu-w** [comment]
The `ModelCheckpoint._save_topk_checkpoint` is run at `on_train_batch_end` in `TrainingEpochLoop.advance`, while `validating` is run at `on_advance_end`
https://github.com/Lightning-AI/lightning/blob/1a3fe395715b2b9eb3c80647b5e0eb490047be8d/src/pytorch_lightning/loops/epoch/training_epoch_loop.py#L233-L256

**Comment by carmocca** [comment]
Tracking a validation metric with `every_n_train_steps` is unsupported. 

This would require one more flag in `ModelCheckpoint` that is equivalent but for validation: `every_n_validation_steps`.
(dev note: unfortunately digging us deeper into https://github.com/Lightning-AI/lightning/issues/4335)

I would not change the loop logic here and treat it as correct.

**Comment by rohitgr7** [comment]
wouldn't this solve your issue?
```py
model_checkpoint = ModelCheckpoint(monitor="valid_loss")
```
since according to `val_check_interval`, validation will happen after every 2 steps and eventually will compute `val_loss`and will call `ModelCheckpoint.on_validation_end`, which will trigger your checkpoint.

FYI: Haven't tried it yet. Just assuming it should not break.

**Comment by carmocca** [comment]
That's right. Thank you @rohitgr7.

We cannot know for sure when will a metric be `log`ged so I don't think there's anything to do here. Closing

**Comment by tshu-w** [comment]
Thanks, @rohitgr7. I initially thought I needed to set checkpoint arguments as well, which is a little confusing.
